src/contract/contract.ts

```diff
@@ -26,165 +26,310 @@ import { node } from '../api'
-      if (this._artifactCache.size >= this.artifactCacheCapacity) {
-        const keyToDelete = this._artifactCache.keys().next().value
-        this._artifactCache.delete(keyToDelete)
+  sourceHasChanged(files: SourceFile[]): boolean {
+    if (files.length != this.files.size) {
+      return true
+    }
+    for (const file of files) {
+      const sourceCodeHash = this.files.get(file.contractPath)
+      if (typeof sourceCodeHash === 'undefined' || sourceCodeHash != file.sourceCodeHash) {
+        return true
       }
-      this._artifactCache.set(contract.codeHash, contract)
     }
+    return false
   }
 
-  constructor(sourceCodeSha256: string, functions: FunctionSig[]) {
-    this.sourceCodeSha256 = sourceCodeSha256
-    this.functions = functions
+  static async from(rootPath: string): Promise<ProjectArtifact | undefined> {
+    const filepath = rootPath + '/' + ProjectArtifact.artifactFileName
+    if (!fs.existsSync(filepath)) {
+      return undefined
+    }
+    const content = await fsPromises.readFile(filepath)
+    const files = new Map(Object.entries<string>(JSON.parse(content.toString())))
+    return new ProjectArtifact(files)
   }
+}
 
-  protected static _artifactsFolder(): string {
-    return './artifacts/'
+export class Project {
+  sourceFiles: SourceFile[]
+  contracts: ArtifactInfo<Contract>[]
+  scripts: ArtifactInfo<Script>[]
+
+  private static instance: Project
+  private static nodeProvider: NodeProvider
+
+  static readonly contractsPath = 'contracts'
+  static readonly artifactsPath = 'artifacts'
+
+  static readonly abstractContractMatcher = new TypedMatcher<SourceType>(
+    '^Abstract Contract [A-Z][a-zA-Z0-9]*',
+    SourceType.AbstractContract
+  )
+  static readonly contractMatcher = new TypedMatcher('^Contract [A-Z][a-zA-Z0-9]*', SourceType.Contract)
+  static readonly interfaceMatcher = new TypedMatcher('^Interface [A-Z][a-zA-Z0-9]* \\{', SourceType.Interface)
+  static readonly scriptMatcher = new TypedMatcher('^TxScript [A-Z][a-zA-Z0-9]*', SourceType.Script)
+  static readonly matchers = [
+    Project.abstractContractMatcher,
+    Project.contractMatcher,
+    Project.interfaceMatcher,
+    Project.scriptMatcher
+  ]
+
+  private constructor(sourceFiles: SourceFile[], contracts: ArtifactInfo<Contract>[], scripts: ArtifactInfo<Script>[]) {
+    this.sourceFiles = sourceFiles
+    this.contracts = contracts
+    this.scripts = scripts
+  }
+
+  static setNodeProvider(nodeProvider: NodeProvider) {
+    Project.nodeProvider = nodeProvider
+  }
+
+  private static getContractPath(path: string): string {
+    return path.startsWith(`./${Project.contractsPath}`)
+      ? path.slice(2)
+      : path.startsWith(Project.contractsPath)
+      ? path
+      : Project.contractsPath + '/' + path
+  }
+
+  contract(path: string): Contract {
+    const contractPath = Project.getContractPath(path)
+    const selected = this.contracts.filter((c) => c.sourceFile.contractPath === contractPath)
```

Comment:
Nit: it seems we could use `find` instead of `filter`

Reply:
Thanks!